trin-execution/src/walkers/trie_walker.rs

```diff
@@ -0,0 +1,378 @@
+/// 1. gossiping the whole state trie
+/// 2. gossiping the forward state diffs
+/// 3. getting stats about the state trie
+pub struct TrieWalker<TrieDB: DB> {
+    is_partial_trie: bool,
+    trie: Arc<TrieDB>,
+    stack: Vec<TrieProof>,
+}
+
+impl<TrieDB: DB> TrieWalker<TrieDB> {
+    pub fn new(root_hash: B256, trie: Arc<TrieDB>) -> anyhow::Result<Self> {
+        let root_value = trie
+            .get(root_hash.as_slice())
+            .expect("Failed to read node from")
+            .expect("Root node not found")
+            .into();
+        let root_proof = TrieProof {
+            path: vec![],
+            proof: vec![root_value],
+        };
+
+        Ok(Self {
+            is_partial_trie: false,
+            trie,
+            stack: vec![root_proof],
+        })
+    }
+
+    pub fn new_partial_trie(root_hash: B256, trie: TrieDB) -> anyhow::Result<Self> {
+        // if the storage root is empty then there is no storage to gossip
+        if root_hash == EMPTY_ROOT_HASH {
+            return Ok(Self {
+                is_partial_trie: true,
+                trie: Arc::new(trie),
+                stack: vec![],
+            });
+        }
+
+        let root_value = match trie
+            .get(root_hash.as_slice())
+            .expect("Failed to read node from")
+        {
+            Some(root_value) => root_value.into(),
+            None => {
+                // The trie db is empty so we can't walk it return an empty iterator
+                return Ok(Self {
+                    is_partial_trie: true,
+                    trie: Arc::new(trie),
+                    stack: vec![],
+                });
+            }
+        };
+
+        let root_proof = TrieProof {
+            path: vec![],
+            proof: vec![root_value],
+        };
+
+        Ok(Self {
+            is_partial_trie: true,
+            trie: Arc::new(trie),
+            stack: vec![root_proof],
+        })
+    }
+
+    fn process_node(
+        &mut self,
+        node: Node,
+        partial_proof: Vec<Bytes>,
+        path: Vec<u8>,
+    ) -> anyhow::Result<()> {
+        // We only need to process hash nodes, because if the node isn't a hash node the leaf is
+        // already embedded in the proof
+        if let Node::Hash(hash) = node {
+            let value_result = self
+                .trie
+                .get(hash.hash.as_slice())
+                .expect("Failed to read node from the database");
+
+            let value = match value_result {
+                Some(value) => value,
+                None => {
+                    // If we are walking a partial trie, some nodes won't be available in the
+                    // database
+                    if self.is_partial_trie {
+                        return Ok(());
+                    }
+                    return Err(anyhow::anyhow!("Node not found in the database"));
+                }
+            };
+            let decoded_node = decode_node(&mut value.as_ref())?;
```

Comment:
nit: I guess we want to do this to make sure that it's valid node. How about something like this:

```rust
// check that node decodes correctly and to correct variant
if matches!(
        decode_node(&mut value.as_ref())?,
        Node::Empty | Node:: Hash(_)) {
    // log warning and return Ok
    // or maybe return Err
}
partial_proof.push(value.into());
self.stack.push(TrieProof { path, proof: partial_proof });
```